docs/framework/api-architecture.md

```diff
@@ -0,0 +1,175 @@
+    }
+    ```
+    
+    ```php
+      function civicrm_api3_survey_get($params) {
+        return _civicrm_api3_basic_get(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+      }
+    ```
+    
+    ```php
+    function civicrm_api3_survey_delete($params) {
+      return _civicrm_api3_basic_delete(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+    }
+    ```
+    
+## Code comment blocks {:#docblock}
+
+-   API comment blocks shall include a description of the action & the
+    input and output params - per the example
+-   API comment blocks shall include references to test generated
+    examples (these are generated by adding
+    `$this->documentMe($params,$result,__FUNCTION__,__FILE__);`
+    to tests
+
+    ```php
+    /**
+     * Create or update a survey
+     *
+     * @param array $params 
+     *   Associative array of property name/value pairs to insert in new 'survey'.
+     * @example SurveyCreate.php Std Create example
+     * @return array api result array
+     * {@getfields survey_create}
+     * @access public
+     */
+    ```
+    
+## `_spec` functions {:#spec}
+
+A "`_spec` function" must exist for every API entity/action pair.
+
+For example is is the [source](https://github.com/civicrm/civicrm-core/blob/1f4ea7262d865c72e8946481fdbfe18f5159da9e/api/v3/Tag.php#L62) of the `_spec` function for the `tag` entity, with the `create` action:
+
+```php
+function _civicrm_api3_tag_create_spec(&$params) {
+  $params['used_for']['api.default'] = 'civicrm_contact';
+  $params['name']['api.required'] = 1;
+  $params['id']['api.aliases'] = array('tag');
+}
+```
+
+If the `_spec` function makes no alterations to `$params`, then each field will behave as defined in the [schema definition](framework/schema-definition/#table-field) for the entity. All of those field-level xml schema tags are also available for use in the API `_spec` function. So for example, you can use `$params['id']['type'] = CRM_Utils_Type::T_INT;` to specify that the `'id'` field must be an integer.
+  
+
+Additionally, the following settings may be applied to each field: 
+
+| Key | Description |
+| -- | -- |
+| `'api.aliases'` | An array of strings which represent alternate names by which this field may be referenced |
+| `'api.default'` | Specifies a default value for the field |
+| `'api.filter'` | ??? |
+| `'api.required'` | Set this to `1` to indicate that the field is required |
+| `'api.return'` | ??? |
+| `'api.unique'` | ??? |
+| `'description'` | A human-readable description of the field which will be displayed in the [API explorer](/api/index.md#api-explorer) |
+| `'options'` | Used to specify an array of values which are considered valid for the field |
+| `'supports_joins'` | ??? |
+| `'FKApiName` | ??? |
+| `'FKClassName'` | Allows the wrapper layer to return useful information if the constraint causes the call to fail. Should be a class name as a string. |
+| `'FKKeyColumn'` | ??? |
```

Comment:
The column on the other table to join to  - this is assumed to be 'id' if not set. But if you join to, for example civicrm_mailing.hash then it will mean the field you are defining joins to civicrm_mailing.hash field